### PR TITLE
Build dependencies out-of-source

### PR DESCRIPTION
**Pull request by joshessman-llnl**

### CODE REVIEW

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -55,7 +55,7 @@ stages:
     - RESOURCES="-t 30 -N 1"
     - echo -e "section_start:$(date +%s):dependencies\r\e[0K
       Build dependencies and generate host-config file (uberenv)"
-    - ${EXEC_PREFIX} ${RESOURCES} scripts/llnl/build_tpls.py --spec=${SPEC} --directory=${CI_PROJECT_NAME}
+    - ${EXEC_PREFIX} ${RESOURCES} scripts/llnl/build_tpls.py --spec=${SPEC} --directory=${CI_BUILDS_DIR}/serac_tpls/${CI_RUNNER_ID}
```
Comment (white238) [inline .gitlab-ci.yml:58]: is this in a directory that will be removed in the normal gitlab manner?

File: .gitlab/build_quartz.yml [hunk 2]
```diff
@@ -43,6 +43,7 @@ release_resources_build_quartz:
 .main_build_with_deps_on_quartz:
   stage: q_build_with_deps
   extends: [.srun_build_with_deps_script, .on_quartz, .main_workflow]
+  needs: []
```
Comment (white238) [inline .gitlab/build_quartz.yml:46]: What did this do?

File: .gitlab-ci.yml (hunk 1 above)
Comment (jamiebramwell) [reply .gitlab-ci.yml:58]: I have to manually cron job clean up our gitlab stuff anyway.

**Review by jamiebramwell — APPROVED** [review]
This looks painful. Thank you for figuring it out!

File: .gitlab-ci.yml (hunk 1 above)
Comment (white238) [reply .gitlab-ci.yml:58]: oh yea.. forgot that awesome feature of gitlab.

**Review by white238 — APPROVED** [review]

File: .gitlab/build_quartz.yml (hunk 2 above)
Comment (joshessman-llnl) [reply .gitlab/build_quartz.yml:46]: It allows the full Quartz build to run in parallel with the Lassen build (which was already the case for the partial PR builds, I just missed it).

Comment (white238) [reply .gitlab/build_quartz.yml:46]: such a clear syntax.

File: .gitlab-ci.yml (hunk 1 above)
Comment (adrienbernede) [reply .gitlab-ci.yml:58]: FYI: putting "CI_RUNNER_ID" does NOT prevent concurrent access to this directory _even in the same pipeline_.
A single runner can handle up to 8 jobs.
At the same time, there are several runners for quartz, so this directory will end up being duplicated.

(note: I like commenting merged PRs ;))

File: .gitlab/build_quartz.yml (hunk 2 above)
Comment (adrienbernede) [reply .gitlab/build_quartz.yml:46]: @white238 that’s sort of a tweak.
What it really says is: "This jobs does not need any other job to start" allowing the job to start right away even if it is in a later stage.
I general, it’s important to keep in mind that once a `needs` keyword is added to a job, it is not restricted to the synchronization barriers between stages anymore.

(note: I like commenting merged PRs ;))

**Pull request merged**